ipn/ipnlocal/local.go

```diff
@@ -3535,6 +3536,77 @@ func (b *LocalBackend) setTCPPortsInterceptedFromNetmapAndPrefsLocked() {
 	b.setTCPPortsIntercepted(handlePorts)
 }
 
+// AddServeConfig receives a ServeConfig and adds TCP/Web mappings
+// to the local serveConfig, returns a read-only view or any errors
+//
+// b.mu must be held.
+func (b *LocalBackend) AddServeConfig(configIn ipn.ServeConfig) (*ipn.ServeConfigView, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	configClone := b.serveConfig.AsStruct()
+
+	// if we have existing and we've also sent in new, assign the new
+	// overwriting the status for existing
+	if configClone.AllowIngress != nil && configIn.AllowIngress != nil {
+		for hp := range configIn.AllowIngress {
+			if configIn.AllowIngress[hp] {
+				configClone.AllowIngress[hp] = true
+			} else {
+				delete(configClone.AllowIngress, hp) // remove false states
+			}
+		}
+	}
+	// if not current AllowIngress mapping, assign all true states
+	if configIn.AllowIngress != nil && configClone.AllowIngress == nil {
+		configIn.AllowIngress = make(map[ipn.HostPort]bool)
+		for hp := range configIn.AllowIngress {
+			if configIn.AllowIngress[hp] { // don't create false states
+				configClone.AllowIngress[hp] = true
+			}
+		}
+	}
+
+	for p := range configIn.TCP {
+		if configClone.TCP == nil {
+			configClone.TCP = make(map[int]*ipn.TCPPortHandler)
+		}
```

Comment:
if you get sick of this, you can use `mak.Set(&m, k, v)` to do these 4 lines in 1 line. We've been using that more & more.